File: packages/contracts/headless/src/hover.rs
//! Hover-intent machine (Tooltip, HoverCard). Mirror of core `hover.ts`.

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HoverState {
    Closed,
    Opening,
    Open,
    Closing,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct HoverContext {
    pub open_delay_ms: f64,
    pub close_delay_ms: f64,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HoverEvent {
    Enter,
    Leave,
    TimerFire,
    Dismiss,
    SetOpen { open: bool },
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum HoverEffect {
    StartTimer { ms: f64 },
    ClearTimer,
    EmitOpenChange { open: bool },
}

fn close_now() -> (HoverState, Vec<HoverEffect>) {
    (
        HoverState::Closed,
        vec![HoverEffect::ClearTimer, HoverEffect::EmitOpenChange { open: false }],
    )
}

pub fn hover_transition(state: HoverState, context: HoverContext, event: HoverEvent) -> (HoverState, Vec<HoverEffect>) {
    match event {
        HoverEvent::Enter => match state {
            HoverState::Open | HoverState::Closing => (HoverState::Open, vec![HoverEffect::ClearTimer]),
            _ => (
                HoverState::Opening,
                vec![HoverEffect::ClearTimer, HoverEffect::StartTimer { ms: context.open_delay_ms }],
            ),
        },
        HoverEvent::Leave => match state {
            HoverState::Closed => (state, vec![]),
            _ if context.close_delay_ms <= 0.0 => close_now(),
            _ => (
                HoverState::Closing,
                vec![HoverEffect::ClearTimer, HoverEffect::StartTimer { ms: context.close_delay_ms }],
            ),
        },
        HoverEvent::TimerFire => match state {
            HoverState::Opening => (HoverState::Open, vec![HoverEffect::EmitOpenChange { open: true }]),
            HoverState::Closing => (HoverState::Closed, vec![HoverEffect::EmitOpenChange { open: false }]),
            _ => (state, vec![]),
        },
        HoverEvent::Dismiss => match state {
            HoverState::Closed => (state, vec![]),
            _ => close_now(),
        },
        HoverEvent::SetOpen { open } => (
            if open { HoverState::Open } else { HoverState::Closed },
            vec![HoverEffect::ClearTimer],
        ),
    }
}

File: packages/core/src/hover.ts
/**
 * Hover-intent behavior machine (Tooltip, HoverCard).
 * Contracts: docs/contracts/components/tooltip.md, hover-card.md,
 * "Behavior Machine" sections.
 *
 * States: closed | opening (open-delay pending) | open | closing
 * (close-delay pending). A zero close delay collapses LEAVE/DISMISS into an
 * immediate close. Immediate and timer-driven closes both emit
 * `emitOpenChange(false)` even when the surface never became visible,
 * matching the pre-machine components.
 */

import type { TransitionResult } from "./machine";
import type {
  HoverState,
  HoverContext,
  HoverEvent,
  HoverEffect,
} from "./generated/machines/hover";

export type {
  HoverState,
  HoverContext,
  HoverEvent,
  HoverEffect,
};

export type HoverResult = TransitionResult<HoverState, HoverContext, HoverEffect>;

function closeNow(context: HoverContext): HoverResult {
  return {
    state: "closed",
    context,
    effects: [{ type: "clearTimer" }, { type: "emitOpenChange", open: false }],
  };
}

export function hoverTransition(
  state: HoverState,
  context: HoverContext,
  event: HoverEvent,
): HoverResult {
  const stay: HoverResult = { state, context, effects: [] };

  switch (event.type) {
    case "ENTER": {
      if (state === "open") {
        // Re-entering while open cancels a pending close.
        return { state: "open", context, effects: [{ type: "clearTimer" }] };
      }

      if (state === "closing") {
        return { state: "open", context, effects: [{ type: "clearTimer" }] };
      }

      return {
        state: "opening",
        context,
        effects: [{ type: "clearTimer" }, { type: "startTimer", ms: context.openDelayMs }],
      };
    }

    case "LEAVE": {
      if (state === "closed") {
        return stay;
      }

      if (context.closeDelayMs <= 0) {
        return closeNow(context);
      }

      return {
        state: "closing",
        context,
        effects: [{ type: "clearTimer" }, { type: "startTimer", ms: context.closeDelayMs }],
      };
    }

    case "TIMER_FIRE": {
      if (state === "opening") {
        return { state: "open", context, effects: [{ type: "emitOpenChange", open: true }] };
      }

      if (state === "closing") {
        return { state: "closed", context, effects: [{ type: "emitOpenChange", open: false }] };
      }

      return stay;
    }

    case "DISMISS": {
      return state === "closed" ? stay : closeNow(context);
    }

    case "SET_OPEN": {
      return { state: event.open ? "open" : "closed", context, effects: [{ type: "clearTimer" }] };
    }
  }
}
